### DISCUSSION TITLE
Axum testing weird problem

### DISCUSSION THREAD

**Discussion by Gabriel-Rosmart** [tokio-rs/axum#1839] [Q&A]
Hi, I'm making a web server that has some routes that return a `Result<Response, AppError>` where `AppError` implements `IntoResponse`, so technically it never returns an error actually, however when testing, endpoints always return 500 status code, even when they work perfectly well (making a request gives correct status code).

Here is the code of AppError:

```
use axum::{
    extract::Json, 
    response::IntoResponse,
    http::StatusCode
};
use validator::ValidationErrors;

pub enum AppError {
    InternalServerError,
    BadRequest(Json<ValidationErrors>),
}

impl From<ValidationErrors> for AppError {
    fn from(value: ValidationErrors) -> Self {
        Self::BadRequest(Json(value))
    }
}

impl IntoResponse for AppError {
    fn into_response(self) -> axum::response::Response {
        match self {
            Self::BadRequest(e) => (StatusCode::BAD_REQUEST, e).into_response(),
            Self::InternalServerError => StatusCode::INTERNAL_SERVER_ERROR.into_response(),
        }
    }
}
```

And here a test (obviously under mod tests):
```
async fn app() -> Router {
        let db = Mongo::init().await.unwrap();

        let routes = Router::new().nest("/auth", auth()).nest("/user", user());

        Router::new().nest("/api", routes).layer(Extension(db))
    }

#[tokio::test]
    async fn login_does_not_accept_empty_body() {
        let app = app().await;
        let uri = "/api/auth/login";

        let response = app
            .oneshot(
                Request::post(uri)
                    .header(http::header::CONTENT_TYPE, "application/json")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    }
```

I'm confused about what to do, don't know if i should be using a fallible service or change something in the test.
This problem came up to me while trying to use the `?` operator for a better readability in my handlers

I have to state that the test works perfectly if i return `Response` in the handler instead of `Result<Response, AppError>`

**Comment by davidpdrsn** [comment]
Can you post a full reproduction script that we can `cargo test` and poke at?

**Comment by Gabriel-Rosmart** [comment]
I've uploaded a repository, so full code is available, which i think should be easier to read. All test are there.
I've also tested endpoints with ThunderClient (they work as expected this way)
The routes mentioned are under `src/api/auth/routes.rs` and `AppError` under `src/errors/server.rs`

Every other important thing either on `main.rs` or `src/api/auth/config.rs`
Link to repo: https://github.com/Gabriel-Rosmart/Deux.git

**Reply by davidpdrsn** [reply]
I'd appreciate it if you made a minimal repro. Makes it easier for us to debug.

**Reply by Gabriel-Rosmart** [reply]
Ok, but what do you mean about a reproduction? Sorry if i don't understand

**Reply by davidpdrsn** [reply]
https://stackoverflow.com/help/minimal-reproducible-example

**Reply by Gabriel-Rosmart** [reply]
Never mind a solved the problem (it was stupid and my fault).
After trying to reproduce the problem, unsuccessfully, i realized that my handlers expected a `Extension<Arc<Database>>` however in my test i added a extension of type `Extension<Database>` while in my main function i used the correct one. After changing it, and wrapping state in an `Arc` everything works as expected.

However this raises a question for me. Trying to share state with the `with_state` function (as the docs suggest) yields an error saying that the function expects a `()` and program does not compile. So i use the `layer` function.

Don't know if this has something to do the the version of axum or not, i'm currently using `0.6.10` version.

**Reply by davidpdrsn** [reply]
Sounds like you're using the wrong state for the router then. See https://docs.rs/axum/0.6.10/axum/routing/struct.Router.html#method.with_state

**Reply by Gabriel-Rosmart** [reply]
Yep, that was it, changed to `State` successfully, thanks for the help.
